Lib/idlelib/hyperparser.py

```diff
@@ -56,7 +56,14 @@ def index2line(index):
                 startatindex = r[1]
             else:
                 startatindex = "1.0"
-            stopatindex = "%d.end" % lno
+            if end_at_eol:
+                stopatindex = "%d.end" % lno
+            else:
+                r = text.tag_nextrange("console", index)
+                if r:
+                    stopatindex = text.index(r[0] + "-1c")
+                else:
+                    stopatindex = "end-1c"
```

Comment:
This `else` is a bit unwieldy.  Since hyperparser has tests, maybe it can be restructured with a high confidence of correctness?
```
        else:
            r = text.tag_prevrange("console", index)
            r = text.tag_prevrange("console", index)
            if r:
            if r:
                startatindex = r[1]
                startatindex = r[1]
            else:
            else:
                startatindex = "1.0"
                startatindex = "1.0"
            stopatindex = "%d.end" % lno
            if end_at_eol:
                stopatindex = "%d.end" % lno
            else:
                r = text.tag_nextrange("console", index)
                if r:
                    stopatindex = text.index(r[0] + "-1c")
                else:
                    stopatindex = "end-1c"
```